docs/proposals/custom-sidecar-config.md

```diff
@@ -0,0 +1,172 @@
+# Custom Container Configuration
+
+## Introduction
+Users and field team have requested access to custom configuration of kubernetes sidecar and init containers.
+See [#3925](https://github.com/kumahq/kuma/issues/3925).
+This intuitively aligns with the ProxyTemplate configuration,
+which provides a free-form configuration of Envoy.
+The current plan is to introduce a new CRD,
+similar to ProxyTemplate,
+which can modify sidecar config utilizing [jsonpatch](https://jsonpatch.com).
+
+## Custom Resource Definition
+The new CRD will be named ContainerTemplate.
+It will allow for customer configuration of
+both sidecar and init containers.
+It will be namespace scoped.
+The webhook will validate that it can only be applied in a namespace where Kuma CP is running.
+The spec will contain an array of jsonpatch strings which describe the modifications to be performed.
+There is no Universal mode equivalent.
+
+### Example
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: ContainerTemplate
+metadata:
+  name: container-template-1
+spec:
+  sidecarTemplate:
+    - op: add
+      path: /securityContext/privileged
+      value: true
+  initTemplate:
+    - op: add
+      path: /securityContext/runAsNonRoot
+      value: true
+    - op: remove
+      path: /securityContext/runAsUser
+
+```
+
+This will change the `securityContext` section of `kuma-sidecar` container from
+
+```yaml
+      securityContext:
+        runAsGroup: 5678
+        runAsUser: 5678
+```
+
+to
+
+
+```yaml
+      securityContext:
+        runAsGroup: 5678
+        runAsUser: 5678
+        priviledged: true
+```
+
+and similarly change the `securityContext` section of the `init` container from
+
+```yaml
+      securityContext:
+        capabilities:
+          add:
+          - NET_ADMIN
+          - NET_RAW
+        runAsGroup: 0
+        runAsUser: 0
+```
+
+to
+
+```yaml
+      securityContext:
+        capabilities:
+          add:
+          - NET_ADMIN
+          - NET_RAW
+        runAsGroup: 0
+        runAsNonRoot: true
+```
+
+## Workload Matching
+A `ContainerTemplate` will be matched to a Pod via an annotation on the workload.
+Each annotation may be an ordered list of `ContainerTemplate` names,
+which will be applied in the order specified.
+
+We will include a configuration option to Kuma CP
+to specify a default list of `ContainerTemplate` patches to apply.
+
+### Example
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  namespace: app-ns
+  name: app-depl
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: app-depl
+  template:
+    metadata:
+      labels:
+        app: app-depl
+      annotations:
+        kuma.io/container-template:
+          - containter-template-1
+          - containter-template-2
+    spec: ...
+```
+
+## Multizone
+
+We will not add any special handling for multizone.
+To patch containers in multizone, the config will need to be pushed to the appropriate zones.
+
+## Error Modes and Validation
```

Comment:
What happens if a user specifies an annotation with a template that doesn't exist?

Reply:
Hmm yes, we will need to bubble that error up somehow. I guess that also raised the question of if we should inject kuma at all in that case. If they need a particular specification for the sidecar/init containers, is it safe to start them with an incomplete patch?

Reply:
UX hat: we shouldn't hard error (just warn and fall back to default)
Security hat: This feature will likely be used for security features so is it 'safe' to just continue on failure? (although there should be another layer of defense to catch this, Gatekeeper, etc...)

Reply:
I guess a log in the CP might be enough.
In the future (when we get down to implement this) an event or a status in the deployment might be the right way to go.
Otherwise I guess a pod validation webhook might be possible no?

Reply:
@lahabana Log, and you think we should just continue with the default?  It seems like most people asking for this so far are trying to get things to work with their platform, which means it will just break for them in the existing way (which is OK IMHO). I wouldn't want to accidentally fail open if there user needs to add some security measure though..

Reply:
Ok for failing I guess you can do defaulting + checking patches exists in `injector.go` where we do sidecar injection anyway.